src/mlpack/core/data/dictionary_encoding.hpp

```diff
@@ -0,0 +1,144 @@
+  */
+  void Reset();
+
+  /**
+  * A fucntion to encode given array of strings using a particular delimiter,
+  * providing custom rule for tokenization.
+  *
+  * For example 
+  * Vector is :
+  * [hello@wow, wow@hello@good] would be encoded using '@' as delimiter as 
+  * [1 2 0, 2 1 3] 
+  * The function paddes 0 to maintain same sizes across all the rows
+  * User may also provide their custom tokenization rule.
+  *
+  * @param strings Vector of strings.
+  * @param output Output Matrix to store encoded results (sp_mat or mat).
+  * @param tokenizer A function that accepts a boost::string_view as
+  *                  an argument and returns a token.
+  * This can either be a function pointer or function object or a lamda
+  * function. Its return value should be a boost::string_view, a token.
+  *
+  * Definiation of function should be of type
+  * boost::string_view fn()(boost::string_view& str)
+  *
+  */
+  template<typename MatType, typename TokenizerType>
+  void Encode(const std::vector<std::string>& strings,
+                  MatType& output, TokenizerType tokenizer);
+
+  /**
+  * A function to encode given array of strings using a particular delimiter,
+  * with custome tokenization.
+  *
+  * For example 
+  * Vector is :
+  * [hello@wow, wow@hello@good] would be encoded using '@' as delimiter as 
+  * [1 2 , 2 1 3] 
+  * The function does not paddes 0 in this case.
+  *
+  * @param strings Vector of strings.
+  * @param output Vector of vectors to store encoded results.
+  * @param tokenizer A function that accepts a boost::string_view as
+  *                  an argument and returns a token.
+  * This can either be a function pointer or function object or a lamda
+  * function. Its return value should be a boost::string_view, a token.
+  *
+  * Definiation of function should be of type
+  * boost::string_view fn()(boost::string_view& str)
+  *
+  */
+  template<typename TokenizerType>
+  void Encode(const std::vector<std::string>& strings,
+            std::vector<std::vector<size_t>>& output, TokenizerType tokenizer);
+
+  //! Modify the stringq.
+  std::deque<std::string>& Stringq() { return stringq; }
+
+  //! Return the stringq.
+  const std::deque<std::string>& Stringq() const { return stringq; }
+
+  //! Return the Mappings
+  const std::unordered_map<boost::string_view, size_t,
+      boost::hash<boost::string_view>>& Mappings() const { return mappings; }
+
+  //! Modify the Mappings.
+  std::unordered_map<boost::string_view, size_t,
+      boost::hash<boost::string_view>>& Mappings() { return mappings; }
+
+  /**
+   * Serialize the class to the given archive.
+   */
+  template<typename Archive>
+  void serialize(Archive& ar, const unsigned int /* version */);
+ private:
+  //! A map which stores information about mapping.
+  std::unordered_map<boost::string_view, size_t,
+      boost::hash<boost::string_view>> mappings;
+  // Vector which hold the string for map's string_view.
+  std::deque<std::string> stringq;
```

Comment:
Hmm.. I don't know how to explain it. I just don't like the name of the variable. Maybe I am too pedantic. But I think each variable name should briefly describe its purpose. Here I don't understand what "q" means. I suggest to use something like `stringDeque`, `stringSet`, `stringCache` or something else. Actually, I am not sure, these examples aren't good enough. Try to find something good.

Reply:
I am not so good at naming things, thought of naming it orginalStrings , how's it..

Reply:
What about `tokens` or `words`?

Reply:
I will use tokens, that is much better.